### PR TITLE
Expand braces in file path

### PR DESCRIPTION
**Pull request by dmpetrov**
Code is Claude AI generated.

It enables ranges in braces `..` like:

```python
.read_storage(f"{my_prefix}/yellow_tripdata_2025-{01..09}.parquet")
```

- numbers: `0..999`
- numbest starting with zeros `01..09`
- letters: `a..f`

## Summary by Sourcery

Extend brace pattern expansion to support numeric (with optional zero-padding and reverse) and alphabetic ranges in storage paths, update documentation examples, and add corresponding unit and functional tests

New Features:
- Support numeric ranges in brace expansion (including zero-padded and reverse sequences)
- Support alphabetic ranges in brace expansion

Enhancements:
- Allow mixing numeric/character ranges with comma-separated lists and multiple brace patterns

Documentation:
- Update expand_brace_pattern docstring to include numeric and character range examples

Tests:
- Add unit tests for numeric range, character range, and complex mixed brace patterns
- Add functional tests in read_storage to verify numeric, zero-padded, and alphabetic range expansions

## Summary by Sourcery

Extend brace expansion in path globbing to support numeric (with optional zero-padding and reverse) and alphabetic ranges alongside existing comma-separated lists.

New Features:
- Support numeric ranges in brace expansion, including zero-padded and reverse sequences
- Support alphabetic ranges in brace expansion

Enhancements:
- Allow combining numeric/character ranges with comma-separated lists and multiple brace patterns

Documentation:
- Update expand_brace_pattern, read_parquet, and read_storage docstrings to document numeric and alphabetic range support

Tests:
- Add unit tests for numeric, zero-padded, reverse, and alphabetic brace expansions and complex mixed patterns
- Add functional tests in read_storage to verify numeric, zero-padded, alphabetic, and combined brace range expansions

### CODE REVIEW

File: src/datachain/lib/dc/parquet.py [hunk 1]
```diff
@@ -26,8 +26,14 @@ def read_parquet(
     """Generate chain from parquet files.
 
     Parameters:
-        path: Storage URI with directory. URI must start with storage prefix such
-            as `s3://`, `gs://`, `az://` or "file:///".
+        path: Storage path(s) or URI(s). Can be a local path or start with a
+            storage prefix like `s3://`, `gs://`, `az://`, `hf://` or "file:///".
```
Comment (dreadatour) [inline src/datachain/lib/dc/parquet.py:30]: May be to also add `http/https` too if we want to merge this PR? https://github.com/iterative/datachain/pull/1331

```suggestion
            storage prefix like `s3://`, `gs://`, `az://`, `hf://`, `https://` or "file:///".
```

or even may be make it as list here 🤔

File: src/datachain/lib/dc/parquet.py [hunk 2]
```diff
@@ -26,8 +26,14 @@ def read_parquet(
...
+            Supports glob patterns:
+              - `*` : wildcard
+              - `**` : recursive wildcard
+              - `?` : single character
+              - `{a,b}` : brace expansion list
+              - `{1..9}` : brace numeric or alphabetic range
```
Comment (dreadatour) [inline src/datachain/lib/dc/parquet.py:36]: 👍🔥

File: src/datachain/lib/dc/parquet.py [hunk 3]
```diff
@@ -43,10 +49,19 @@ def read_parquet(
...
+        All files from a directory:
         ```py
-        import datachain as dc
-        dc.read_parquet("s3://mybucket/dir")
+        dc.read_parquet("s3://mybucket/dir/")
+        ```
+
+        Only parquet files from a directory, and all it's subdirectories:
+        ```py
+        dc.read_parquet("s3://mybucket/dir/**/*.parquet")
```
Comment (dreadatour) [inline src/datachain/lib/dc/parquet.py:59]: Hm... Just a question: will it works with `http/https` schema after [this PR](https://github.com/iterative/datachain/pull/1331) will be merged? 🤔 I guess it is now, — should we handle this case then?

Like for some schemas we only can use some glob patterns. E.g. for `http/https` we can only use:
- `{a,b}` : brace expansion list
- `{1..9}` : brace numeric or alphabetic range
But we can not use:
- `*` : wildcard
- `**` : recursive wildcard
- `?` : single character
What do you think? Should we handle this cases and/or notice that in docs? 🤔

Just random idea: may be to add some attributes to FS clients, like: `supports_wildcards`, `supports_brace_expansion`, etc.

File: src/datachain/lib/dc/storage_pattern.py [hunk 4]
```diff
@@ -12,61 +12,37 @@ def validate_cloud_bucket_name(uri: str) -> None:
...
-    Args:
-        uri: URI to validate
-
     Raises:
         ValueError: If a cloud storage bucket name contains glob patterns
     """
     if not is_cloud_uri(uri):
         return
 
-    # Extract bucket name (everything between :// and first /)
```
Comment (dreadatour) [inline src/datachain/lib/dc/storage_pattern.py:24]: 👍

File: src/datachain/lib/dc/parquet.py (hunk 3 above)
Comment (dmpetrov) [reply src/datachain/lib/dc/parquet.py:59]: That's right - braces could be used in http while the wildcards and `?` cannot. It's due to dir logic - http usually doesn't support directory reading and even it supports it's much more involved http implementation.

**Comment by dmpetrov** [comment]
> Will it be the subject of your next PR? 😬

🙂  I hope no, I was planning to support only unix CLI patterns, not regexp
